closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -72,526 +81,928 @@
+                // spec mentioned that only return recovered data after tag
+                // is successfully verified
+                initBuffer(inLen);
+                ibuffer.write(in, inOfs, inLen);
+            }
+            return 0;
+        }
+
+        @Override
+        public int doUpdate(ByteBuffer src, ByteBuffer dst)
+            throws ShortBufferException {
+            int bLen = getBufferedLength();
+            checkDataLength(bLen, src.remaining());
+            if (src.remaining() > 0) {
+                // store internally until doFinal is called because
+                // spec mentioned that only return recovered data after tag
+                // is successfully verified
+                initBuffer(src.remaining());
+                byte[] b = new byte[src.remaining()];
+                src.get(b);
+                // remainder offset is based on original buffer length
+                try {
+                    ibuffer.write(b);
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+            return 0;
+        }
+
+        /**
+         * Performs decryption operation for the last time.
+         *
+         * <p>NOTE: For cipher feedback modes which does not perform
+         * special handling for the last few blocks, this is essentially
+         * the same as <code>encrypt(...)</code>. Given most modes do
+         * not do special handling, the default impl for this method is
+         * to simply call <code>decrypt(...)</code>.
+         *
+         * @param in the input buffer with the data to be decrypted
+         * @param inOfs the offset in <code>cipher</code>
+         * @param len the length of the input data
+         * @param out the buffer for the decryption result
+         * @param outOfs the offset in <code>plain</code>
+         * @return the number of bytes placed into the <code>out</code> buffer
+         */
+        @Override
+        public int doFinal(byte[] in, int inOfs, int inLen, byte[] out,
+            int outOfs) throws IllegalBlockSizeException, AEADBadTagException,
+            ShortBufferException {
+            if (inLen < 0) {
+                throw new ProviderException("Input length is negative");
+            }
+            int bLen = getBufferedLength();
+            if (inLen < tagLenBytes - bLen) {
+                throw new AEADBadTagException("Input too short - need tag");
+            }
+            if (inLen > MAX_BUF_SIZE - bLen) {
+                throw new ProviderException("SunJCE provider only supports "
+                    + "a positive input size up to " + MAX_BUF_SIZE + " bytes");
+            }
+            if (out.length - outOfs < inLen  + bLen - tagLenBytes) {
+                throw new ShortBufferException("Output buffer too small");
+            }
+            byte[] aad = ((aadBuffer == null || aadBuffer.size() == 0) ?
```

Comment:
More parentheses in these conditions, but not the outermost on line 923.

Reply:
done, thank you